src/pkg/oci/pull.go

```diff
@@ -0,0 +1,241 @@
+//
+// layersToPull is an optional parameter that allows the caller to specify which layers to pull.
+//
+// The following layers will ALWAYS be pulled if they exist:
+//   - zarf.yaml
+//   - checksums.txt
+//   - zarf.yaml.sig
+func (o *OrasRemote) PullPackage(destinationDir string, concurrency int, layersToPull ...ocispec.Descriptor) error {
+	isPartialPull := len(layersToPull) > 0
+	ref := o.Reference
+	message.Debugf("Pulling %s", ref.String())
+	message.Infof("Pulling Zarf package from %s", ref)
+
+	manifest, err := o.FetchRoot()
+	if err != nil {
+		return err
+	}
+
+	estimatedBytes := int64(0)
+	if isPartialPull {
+		for _, path := range AlwaysPull {
+			desc := manifest.Locate(path)
+			layersToPull = append(layersToPull, desc)
+			estimatedBytes += desc.Size
+		}
+		for _, desc := range layersToPull {
+			path := desc.Annotations[ocispec.AnnotationTitle]
+			if !utils.SliceContains(AlwaysPull, path) {
+				estimatedBytes += desc.Size
+			}
+		}
+	} else {
+		estimatedBytes = manifest.SumLayersSize()
+	}
+	estimatedBytes += manifest.Config.Size
+
+	dst, err := file.New(destinationDir)
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+
+	copyOpts := o.CopyOpts
+	copyOpts.Concurrency = concurrency
+	if isPartialPull {
+		paths := []string{}
+		for _, layer := range layersToPull {
+			paths = append(paths, layer.Annotations[ocispec.AnnotationTitle])
+		}
+		copyOpts.FindSuccessors = func(ctx context.Context, fetcher content.Fetcher, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+			nodes, err := content.Successors(ctx, fetcher, desc)
+			if err != nil {
+				return nil, err
+			}
+			var ret []ocispec.Descriptor
+			for _, node := range nodes {
+				if utils.SliceContains(paths, node.Annotations[ocispec.AnnotationTitle]) {
+					ret = append(ret, node)
+				}
+			}
+			return ret, nil
+		}
+	}
+
+	// Create a thread to update a progress bar as we save the package to disk
+	doneSaving := make(chan int)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go utils.RenderProgressBarForLocalDirWrite(destinationDir, estimatedBytes, &wg, doneSaving, "Pulling Zarf package data")
+	_, err = oras.Copy(o.Context, o.Repository, ref.String(), dst, ref.String(), copyOpts)
+	if err != nil {
+		return err
+	}
+
+	// Send a signal to the progress bar that we're done and wait for it to finish
+	doneSaving <- 1
+	wg.Wait()
+
+	message.Debugf("Pulled %s", ref.String())
+	message.Successf("Pulled %s", ref.String())
+
+	layersToCheck := []string{}
+	if isPartialPull {
+		for _, layer := range layersToPull {
+			if layer.Annotations[ocispec.AnnotationTitle] != "" {
+				layersToCheck = append(layersToCheck, layer.Annotations[ocispec.AnnotationTitle])
+			}
+		}
+	}
+	return utils.ValidatePackageChecksums(destinationDir, layersToCheck)
+}
+
+// PullFileLayer pulls a file layer from the remote repository and saves it to `destinationDir/annotationTitle`.
+func (o *OrasRemote) PullFileLayer(desc ocispec.Descriptor, destinationDir string) error {
```

Comment:
```suggestion
func (o *OrasRemote) PullLayer(desc ocispec.Descriptor, destinationDir string) error {
```

This could probably be PullLayer to be cleaner since File has a different meaning in Zarf.